drivers/interrupt_controller/intc_plic.c

```diff
@@ -93,18 +95,60 @@ static inline uint32_t get_plic_enabled_size(const struct device *dev)
 	return local_irq_to_reg_index(config->num_irqs) + 1;
 }
 
+static inline uint16_t get_first_context(uint16_t hartid)
+{
+	return hartid == 0 ? 0 : (hartid * 2) - 1;
+}
+
+static inline mem_addr_t get_context_en_addr(const struct device *dev, uint16_t cpu_num)
+{
+	const struct plic_config *config = dev->config;
+	int hartid;
+	/*
+	 * We want to return the irq_en address for the context of given hart.
+	 * If hartid is 0, we return the devices irq_en property, job done. if it is
+	 * greater than zero, we assume that there are two context's associated with
+	 * each hart: M mode enable, followed by S mode enable. We return the M mode
+	 * enable address.
+	 */
+#if CONFIG_SMP
+	hartid = _kernel.cpus[cpu_num].arch.hartid;
+#else
+	hartid = arch_proc_id();
+#endif
+	return  config->irq_en + get_first_context(hartid) * PLIC_REG_IRQ_EN_SIZE;
+}
+
 static inline mem_addr_t get_claim_complete_addr(const struct device *dev)
 {
 	const struct plic_config *config = dev->config;
 
-	return config->reg + PLIC_REG_REGS_CLAIM_COMPLETE_OFFSET;
+	/*
+	 * We want to return the claim complete addr for the hart's context.
+	 * We are making a few assumptions here:
+	 * 1. for hart 0, return the first context claim complete.
+	 * 2. for any other hart, we assume they have two privileged mode contexts
+	 * which are contiguous, where the m mode context is first.
+	 * We return the m mode context.
+	 */
+
+	return config->reg + get_first_context(arch_proc_id()) * PLIC_REG_REGS_SIZE +
```

Comment:
why aren't we passing `hartid` into `get_first_context`, or why is this different from the implementation in `get_threshold_priority_addr`?

Reply:
@ycsin
We need to get the the threshold_priority_addr for all harts running Zephyr:
```
struct _cpu_arch {
#ifdef CONFIG_USERSPACE
	unsigned long user_exc_sp;
	unsigned long user_exc_tmp0;
	unsigned long user_exc_tmp1;
#endif
#if defined(CONFIG_SMP) || (CONFIG_MP_MAX_NUM_CPUS > 1)
	unsigned long hartid;
	bool online;
```
so arch.hartid is only available if ```CONFIG_SMP=y```

When it comes to the writing to the claim complete register, we want to only write to the context for the hart that serviced the interrupt. They reason I used the arch_proc_id() inline is because in the future, Zephyr may run in Supervisor mode. in that scenario, you cannot read mhartid, so that logic will need to be added to that inline (in one place only).